### DISCUSSION TITLE
Lazily finding first item in an array of TaskEithers

### DISCUSSION THREAD

**Discussion by markgaylard** [gcanti/fp-ts#1670] [Q&A]
Hi. Given the following functions:

```typescript
declare const getNumber = (id: string): TaskEither<E, O.Option<number>>;

declare const getFirstSuccess = (ids: string[]): TaskEither<E, O.Option<number>>;
```

I'm trying to work out how to (lazily) find the first `O.some` result. My initial idea was to implement `getFirstSuccess` with something like `pipe(ids, A.findFirstMap(getNumber))` but that doesn't work because `findFirstMap` wants a function that returns an Option. I think I can make it work by using `A.map(getNumber)` then `TE.sequenceArray` then folding the result and doing a `A.findFirst` on the result, but ideally I'd like this to only call `getNumber` until i get the first non-none result.

Is this possible? If so, how? :) 

Thanks.

**Comment by imcotton** [comment]
Could use `altAll` to perform short-circuit evaluation throughout array:

```ts
import {
    alt,
    function as F,
    option as O,
    readonlyArray as A,
    taskEither as TE,
} from 'fp-ts';



declare const getNumber: (id: string) => TE.TaskEither<Error, O.Option<number>>;



const getFirstSuccess: (ids: string[]) => TE.TaskEither<Error, number> = F.flow(

    A.map(F.flow(
        getNumber,
        TE.chain(TE.fromOption(() => new Error('no number'))),
    )),

    alt.altAll (TE.Alt) (TE.left(new Error('empty array'))),

);
```

Note that in order to play well, you'd want to lift and flat nested `Option` into `Either`.
